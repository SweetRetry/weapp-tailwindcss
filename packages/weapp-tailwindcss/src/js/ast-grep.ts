import { logger } from '../logger'

export async function getAstGrep() {
  try {
    const { js } = await import('@ast-grep/napi')
    return js
  }
  catch (error) {
    logger.warn('请先安装 `@ast-grep/napi` , 安装完成后再尝试运行！')
    throw error
  }
}
