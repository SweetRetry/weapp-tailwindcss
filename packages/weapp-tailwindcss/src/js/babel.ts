import type { NodePath } from '@babel/traverse'
import type { Node } from '@babel/types'

export function isEvalPath(p: NodePath<Node>) {
  if (p.isCallExpression()) {
    const calleePath = p.get('callee')
    return calleePath.isIdentifier() && calleePath.node.name === 'eval'
  }
  return false
}
