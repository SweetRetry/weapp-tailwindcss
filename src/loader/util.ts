import fs from 'fs'
import path from 'path'

export function mkfileSync (filename: string, content: string) {
  try {
    const dirname = path.dirname(filename)
    if (!fs.existsSync(dirname)) {
      fs.mkdirSync(dirname, {
        recursive: true
      })
    }
    fs.writeFileSync(filename, content, 'utf-8')
  } catch (error) {
    console.error(error)
  }
}
