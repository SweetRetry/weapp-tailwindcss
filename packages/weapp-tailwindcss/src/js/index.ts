import type { CreateJsHandlerOptions, IJsHandlerOptions } from '../types'
import { defuOverrideArray } from '../utils'
import { jsHandlerAsync } from './ast-grep'
import { jsHandler } from './babel'

export {
  jsHandler,
  jsHandlerAsync,
}

export function createJsHandler(options: CreateJsHandlerOptions) {
  const {
    mangleContext,
    arbitraryValues,
    escapeMap,
    jsPreserveClass,
    generateMap,
    jsAstTool,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
  } = options

  function handler(rawSource: string, classNameSet: Set<string>, options?: CreateJsHandlerOptions) {
    const opts = defuOverrideArray<IJsHandlerOptions, IJsHandlerOptions[]>(options as IJsHandlerOptions, {
      classNameSet,
      escapeMap,
      arbitraryValues,
      mangleContext,
      jsPreserveClass,
      generateMap,
      jsAstTool,
      babelParserOptions,
      ignoreCallExpressionIdentifiers,
      ignoreTaggedTemplateExpressionIdentifiers,
    })
    if (opts.jsAstTool === 'ast-grep') {
      return jsHandlerAsync(rawSource, opts)
    }
    return jsHandler(rawSource, opts)
  }

  function sync(rawSource: string, classNameSet: Set<string>, options?: CreateJsHandlerOptions) {
    const opts = defuOverrideArray<IJsHandlerOptions, IJsHandlerOptions[]>(options as IJsHandlerOptions, {
      classNameSet,
      escapeMap,
      arbitraryValues,
      mangleContext,
      jsPreserveClass,
      generateMap,
      jsAstTool,
      babelParserOptions,
      ignoreCallExpressionIdentifiers,
      ignoreTaggedTemplateExpressionIdentifiers,
    })
    return jsHandler(rawSource, opts)
  }

  handler.sync = sync

  return handler
}
