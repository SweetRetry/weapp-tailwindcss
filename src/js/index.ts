import { parse, traverse, generate } from '@/babel'
import type { TraverseOptions, IJsHandlerOptions } from '@/types'
import type { Node, StringLiteral, TemplateElement } from '@babel/types'
import { replaceWxml } from '@/wxml/shared'
import { escapeStringRegexp } from '@/reg'
import { splitCode } from '@/extractors/split'
import { isProd } from '@/env'

export function handleValue(str: string, node: StringLiteral | TemplateElement, options: IJsHandlerOptions) {
  const set = options.classNameSet
  const escapeEntries = options.escapeEntries
  const arr = splitCode(str) // .split(/\s/).filter((x) => x) // splitCode(n.value) // .split(/\s/).filter((x) => x)
  let rawStr = str
  for (let i = 0; i < arr.length; i++) {
    const v = arr[i]
    if (set.has(v)) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag = node.leadingComments.findIndex((x) => x.value.includes('weapp-tw') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        rawStr = rawStr.replace(
          new RegExp(escapeStringRegexp(v), 'g'),
          replaceWxml(v, {
            escapeEntries
          })
        )
      }
    }
  }
  return rawStr
}

export function jsHandler(rawSource: string, options: IJsHandlerOptions) {
  const ast = parse(rawSource)
  // 这样搞会把原先所有的 children 含有相关的 也都转义了
  const topt: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        const n = p.node
        n.value = handleValue(n.value, n, options)
      }
      // exit(p) {}
    },
    TemplateElement: {
      enter(p) {
        const n = p.node
        n.value.raw = handleValue(n.value.raw, n, options)
      }
    },
    noScope: true
  }

  traverse(ast, topt)

  return generate(ast, {
    minified: options.minifiedJs ?? isProd()
  })
}

export function createjsHandler(options: Omit<IJsHandlerOptions, 'classNameSet'>) {
  return (rawSource: string, set: Set<string>) => {
    return jsHandler(rawSource, {
      classNameSet: set,
      escapeEntries: options.escapeEntries,
      minifiedJs: options.minifiedJs
    })
  }
}
