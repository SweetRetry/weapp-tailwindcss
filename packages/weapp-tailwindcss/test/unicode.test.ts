import { createGetCase, unicodeCasePath } from './util'
import { decodeUnicode } from '@/js/handlers'
import { jsHandler, jsHandlerAsync } from '@/js'

const getCase = createGetCase(unicodeCasePath)

describe('unicode', () => {
  it('unicode case 0', () => {
    const testCase = '\u0009\u000A\u000B\u000C\u000D\u0020\u00A0\u1680\u2000\u2001\u2002'
      + '\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028\u2029\uFEFF'
    expect(decodeUnicode(testCase)).toMatchSnapshot()
  })

  it('unicode case 0 fs', async () => {
    const testCase = await getCase('case0.js')

    const { code } = await jsHandlerAsync(testCase, {
      jsAstTool: 'ast-grep',
    })
    expect(code).toMatchSnapshot()
  })

  it('unicode case 1 fs', async () => {
    const testCase = await getCase('case0.js')

    const { code } = await jsHandler(testCase, {

    })
    expect(code).toMatchSnapshot()
  })
})
