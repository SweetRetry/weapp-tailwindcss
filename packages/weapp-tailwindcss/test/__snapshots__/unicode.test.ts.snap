// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unicode > unicode case 0 1`] = `
"	

                　  ﻿"
`;

exports[`unicode > unicode case 0 fs 1`] = `
"const testCase = '\\u0009\\u000a\\u000b\\u000c\\u000d      '
      + '          　  ﻿'"
`;

exports[`unicode > unicode case 1 fs 1`] = `
"const testCase = '\\u0009\\u000A\\u000B\\u000C\\u000D\\u0020\\u00A0\\u1680\\u2000\\u2001\\u2002'
      + '\\u2003\\u2004\\u2005\\u2006\\u2007\\u2008\\u2009\\u200A\\u202F\\u205F\\u3000\\u2028\\u2029\\uFEFF'"
`;

exports[`unicode > unicode case 2 fs 1`] = `"const x = "afterccontent-_qu6211u77e5u9053u6211u5fc3bu6c38u605212we_dsq_""`;

exports[`unicode > unicode case 2 fs 1 1`] = `"const x = "after:content-['我知道我心,永恒12we_ds']""`;
